manager/controlapi/node.go

```diff
@@ -248,6 +248,26 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 	}, nil
 }
 
+func removeNodeAttachments(tx store.Tx, nodeID string) error {
+	// remove the node's attached containers. if we don't do this, the
+	// network these attachments are connected to will never be removeable
+	tasks, err := store.FindTasks(tx, store.ByNodeID(nodeID))
+	if err != nil {
+		return err
+	}
+	for _, task := range tasks {
+		// if the task is an attachment, then we just delete it. the allocator
+		// will do the heavy lifting. basically, GetAttachment will return the
+		// attachment if that's the kind of runtime, or nil if it's not.
+		if task.Spec.GetAttachment() != nil {
```

Comment:
Spoke with @dperny offline but capturing here for everyone else: I guess it makes sense to enforce a model where the task reaper is the only place where tasks are deleted. We could mark these tasks as orphaned and let them be cleaned out by the reaper, with the caveat that the network attachments are removable for orphaned tasks.

Reply:
I agree with marking the tasks `Orphaned` when the associated node is deleted. I think that's exactly the right behavior.

This should be done in the orchestrators, which is what controls task lifecycle. For example, in the replicated orchestrator:

```go
func (r *Orchestrator) handleTaskEvent(ctx context.Context, event events.Event) {
        switch v := event.(type) {
        case api.EventDeleteNode:
                r.restartTasksByNodeID(ctx, v.Node.ID)
```

We would want to modify this code to set the old task's state (not desired state) to `Orphaned`. The code here might be a little hard to follow, because the tasks get added to a map that's eventually processed as a batch, with those tasks getting passed to `Restart`. We could potentially have a separate map for tasks that need to become orphaned, and pass them to the restart manager in the same way, but then set the state to `Orphaned` after calling `Restart`.

The global orchestrator would need similar changes.

For network attachment tasks, I'm not entirely sure what the best way is. We could create a simple orchestrator for those that just watches for node deletion events. I think that's the cleanest way, but as a simpler kludge we could handle this in the network allocator.

I don't think any changes in the task reaper or network allocator are necessary, because once a task's state is set to `Orphaned`, its network resources are supposed to be freed. But it's definitely worth confirming that this works as expected after the node has been deleted.

Reply:
I see that when we remove a node, we  delete all tasks for that node from the store. What is the reason to not keep the history around in that case ? Since the service may still be around, doesn't it make sense to keep the task history from that node around ? @aaronlehmann

Reply:
It was done for global service tasks because otherwise these tasks would stay in the store forever. The task reaper keeps a certain number per node, so there's no provision for removing all the tasks from dead nodes. I'm not sure I see a better way than deleting it immediately, because with the node no longer in the system, we'll never know when the task has shut truly shut down, and is finally safe to delete. Possibly we could use the orphaned state, as discussed above, though if we wanted to be really careful, we would set that state after a delay passes, like we do for unresponsive nodes.

For reference, here's the commit that made the change: 56463e4f79cec67bac7e2cd80cd01bd8e008b6d2